src/streaming-ingestion/redshift/streaming-ingestion-associate-role.ts

```diff
@@ -0,0 +1,202 @@
+import { AssociateIAMRoleToRedshift, ExistingRedshiftServerlessProps, ProvisionedRedshiftProps } from '../../analytics/private/model';
+import { createLambdaRole } from '../../common/lambda';
+import { POWERTOOLS_ENVS } from '../../common/powertools';
+import { SolutionNodejsFunction } from '../../private/function';
+
+export interface StreamingIngestionAssociateRoleProps {
+  readonly projectId: string;
+  readonly appIds: string;
+  readonly serverlessRedshift?: ExistingRedshiftServerlessProps;
+  readonly provisionedRedshift?: ProvisionedRedshiftProps;
+  readonly databaseName: string;
+  readonly dataAPIRole: IRole;
+  readonly streamingIngestionRole: IRole;
+  readonly streamingIngestionProps: StreamingIngestionProps;
+}
+
+export class StreamingIngestionAssociateRole extends Construct {
+
+  public readonly crForModifyClusterIAMRoles: CustomResource;
+
+  constructor(scope: Construct, id: string, props: StreamingIngestionAssociateRoleProps) {
+    super(scope, id);
+
+    // custom resource to associate the IAM role to redshift cluster
+    this.crForModifyClusterIAMRoles = this.createCustomResourceAssociateIAMRole(props, props.streamingIngestionRole);
+  }
+
+  private createCustomResourceAssociateIAMRole(props: StreamingIngestionAssociateRoleProps, streamingIngestionRole: IRole): CustomResource {
+    const fn = new SolutionNodejsFunction(this, 'StreamingIngestionAssociateIAMRoleToRedshiftFn', {
+      runtime: Runtime.NODEJS_18_X,
+      entry: join(
+        __dirname + '/../../analytics/lambdas/custom-resource',
+        'redshift-associate-iam-role.ts',
+      ),
+      handler: 'handler',
+      memorySize: 256,
+      reservedConcurrentExecutions: 1,
+      timeout: Duration.minutes(5),
+      logRetention: RetentionDays.ONE_WEEK,
+      role: createLambdaRole(this, 'StreamingIngestionAssociateIAMRoleFnRole', false, [
+        new PolicyStatement({
+          actions: [
+            'iam:PassRole',
+          ],
+          resources: ['*'], // have to use wildcard for keeping existing associated roles
+        }),
+      ]),
+      environment: {
+        ...POWERTOOLS_ENVS,
+      },
+    });
+
+    const provider = new Provider(
+      this,
+      'StreamingIngestionRedshiftAssociateIAMRoleCustomResourceProvider',
+      {
+        onEventHandler: fn,
+        logRetention: RetentionDays.FIVE_DAYS,
+      },
+    );
+
+    const customProps: AssociateIAMRoleToRedshift = {
+      roleArn: streamingIngestionRole.roleArn,
+      serverlessRedshiftProps: props.serverlessRedshift,
+      provisionedRedshiftProps: props.provisionedRedshift,
+    };
+
+    const cr = new CustomResource(this, 'StreamingIngestionRedshiftAssociateIAMRoleCustomResource', {
+      serviceToken: provider.serviceToken,
+      properties: customProps,
+    });
+
+    if (props.serverlessRedshift) {
```

Comment:
Looks like they are duplicated code with the data modeling(Redshift), let's extract them as method to share the code.

Reply:
done